### DISCUSSION TITLE
プレーヤー側の朦朧に関する新しい仕様

### DISCUSSION THREAD

**Discussion by Hourier** [hengband/hengband#1492] [Ideas]
# まえがき
表題の件について、開発メンバーだけではなく、ユーザ一般からも意見を募ることを前提としつつ仕様を固めたいと考えています
これにより直接チケットを作るのではなくDiscussion扱いとしました

# 仕様

## 朦朧蓄積ランクの決定
＠が受ける直接打撃の実ダメージに応じてランクを決める (※1)
値は0～8までの9通りとする (計算式の関係上9以上になる可能性もあるが、その時は後続処理で8とみなす)

直接打撃に関して、ランクは以下のように定義する：
- 一定条件がtrue：ランク0
- 実ダメージ21～35：ランク1
- 実ダメージ36～50：ランク2
- 実ダメージ51～65：ランク3
- 実ダメージ66～80：ランク4
- 実ダメージ81～95：ランク5
- 実ダメージ96～110：ランク6
- 実ダメージ111～255：ランク7
- 実ダメージ256～：ランク8

以下は別判定 (後述する朦朧蓄積値を直接加算している。新仕様でどうするかは未定)である
記載漏れがあれば情報求む
- 轟音・プラズマ・フォース・重力のブレス
- 極寒の矢
- ウォーター・ボール
- ウォーター＆プラズマ・ボルト
- ロケット
- 打撃属性：朦朧
- 爆発の薬
- 箱トラップ：邪悪なルーン
- 超能力 (失敗時)

ランク0になる条件は、以下のいずれか1つが満たされた場合とする
- 実ダメージ < ダイス目の最大値 (XdY に対してX*Y) * 0.95
- 実ダメージ：0～20 (4d5の体当たりは、打撃属性がSUPERHURT以外ならば事実上フレーバー化)

この後、ランク上昇/下降判定を行う
ランク0ならば上昇も下降もせずスキップする
- 痛恨の一撃 (判定基準：実ダメージ > 最大ダイス目)、かつ現在のランクが1～6ならば、ランクを1だけ上げる
- 2%の確率でランクを1だけ上げる (※2)
- 腕力＋器用＋耐久＞80 ならばランクを1だけ下げる
  ここで、例えば18/109 はfloor(18 + 109/10) == 28とカウントする

## 朦朧蓄積値の決定
朦朧蓄積ランクに応じて、朦朧蓄積値を決める
大雑把に言って、これが「頭がはっきりするまでのターン数」となる
- ランク0：0
- ランク1：1～10
- ランク2：11～20
- ランク3：21～30
- ランク4：31～40
- ランク5：41～50
- ランク6：51～60
- ランク7：61～70
- ランク8：71～80

## 朦朧ランクの決定
現時点での蓄積値＋上記のダメージで受けた蓄積値に応じて、朦朧ランクを決める
- 朦朧蓄積値1～50：やや朦朧 ← 追加
- 朦朧蓄積値50～100：朦朧
- 朦朧蓄積値101～150：ひどく朦朧
- 朦朧蓄積値151～200：意識不明瞭
- 朦朧蓄積値201～：昏倒 ← 追加

## 朦朧ペナルティ
以下の事項にペナルティを受ける (他にもあるかも、情報募集)
- 攻撃をミスする確率
- 射撃をミスする確率
- 攻撃を避ける確率
- 射撃を避ける確率 (← 斬鉄剣のみ特別実装で、これを拡張。その他は元々必中なので、この作業が完了したら拡張したい)
- 魔法を唱える確率 (種族レイシャル、クラスレイシャル、突然変異も含)
- 各種アイテムを使える確率
  (ひどく朦朧としていたら巻物なんて読めないであろう。薬を飲もうにも手が震えていよう。他検討中)
- 落馬する確率
- 移動できる確率
  (上り階段含。下りは転げ落ちるダメージ：10d5と引き換えに降りられる)

ペナルティは一律で以下の通りとする
- やや朦朧：10%
- 朦朧：20%
- ひどく朦朧：30%
- 意識不明瞭：50%
- 昏倒：100%

## 朦朧からの回復
現行通りとする
1ターンに最低1回復し、かつ耐久値によりボーナスがある (合計で最大10)

# バグ詳細
※1
Alpha37までは、朦朧蓄積ランク1～2 (実ダメージ12～24)の値にほぼ意味がなかった
(ランク0は朦朧蓄積なしなので除外)
これが一撃で意識不明瞭になる原因の1つだった
要するに「想定値よりかなりの高確率で2ランク上の朦朧値が蓄積される」が一撃で意識不明瞭になる原因
また、ダメージの閾値も相当に低い (これはZangbandでインフレしたからかもしれない)
十分バグと呼ぶに相応しい
なおバグの作り込みは約30年前なので、もう詳細を追うことは不可能
他には、例えばウォーター・ボール等も、轟音耐性がない状態で3発受けると意識不明瞭になる可能性がある

※2
Alpha37までは、低確率で無制限に朦朧蓄積ランクが上がっていった
即ち、理論上20d1でも一撃で意識不明瞭になる可能性があった
これを緩和する

**Comment by Hourier** [comment]
Discordで出た意見
- ひどく朦朧 → 昏倒 は認めるが、意識不明瞭 → 昏倒 は認めず意識不明瞭のままとする
  → ひどく朦朧 から蓄積ランク6以上のダメージを受ける状態ならば、朦朧以前にHPが危ない
  これは妥当な意見と思えるので採用
- 能力値が弱ければ弱いほどランクが上がる方針かと思っていた
  → これもアリ。例えば「腕力＋器用＋耐久＜40 ならばランクを1上げる」等
  後から調整は可能なので、まずは緩和のみ実施する

**Comment by Hourier** [comment]
備考
- 朦朧蓄積ランクの閾値は現行通りの95%でなくともいいかもしれない。例えば90%とか
- 階段から転げ落ちる処理は朦朧ランクに応じてダメージを可変にしてもいいかも。5d5、10d5、15d5、……
- 階段から転げ落ちて死んだ時の処理も要追加：「階段に殺された」は何か違う気もする。後で考える

**Comment by HiddenNameZ** [comment]
やや朦朧状態でも巻物や薬使用に失敗するのは連続失敗の恐れがあり、朦朧を発生させるモンスター全てに対して忌避行動を取らせるようになると思います ただ、魔法書失敗率上昇から類推するに巻物失敗は面白いとも思います
自分の脳内イメージ的には朦朧としていても薬のガブ飲みくらいはできると思うのでひどく朦朧とか意識不明瞭とかで薬使用に失敗率が発生したほうが良いと思います

**Comment by chikatei** [comment]
薬と巻物は確実な危機回避手段のままであってほしいので、朦朧時でも確実に成功してほしいです。ステータスが一定値以上になれば失敗を防げる、とかもありかもしれませんが、死を避けるために初心者が覚えるべきルールが増えるのはちょっと微妙な気がします。
魔道具の失敗率が上がるのなら大勢に影響ないと思います。

**Comment by Hourier** [comment]
ご意見ありがとうございます
回答の前に現行仕様について補足します

## 朦朧蓄積ランクの決定 (抜粋)

- ダメージが46以上ならばランク6確定
- 上記に加え、痛恨の一撃 or 2%の確率でランク7へ上昇
  例：23d2の最大ダイス目 (46) かつ 痛恨発動(2倍) かつ AC125 (半分) → 実ダメージ46
- その他、確率で無制限にランク上昇
  (最悪パターン)ダメージが20 かつ 21/100 * 16/1億 の確率でランク7

## 朦朧蓄積値の決定 (抜粋)

- ランク6：80固定
- ランク7：150固定

## 朦朧ランクの決定

- 蓄積値1～50：朦朧
- 蓄積値51～100：ひどく朦朧
- 蓄積値101～：意識不明瞭

## ペナルティ (抜粋)
- 朦朧：魔法/レイシャル失敗率15%
- ひどく朦朧：同25%
- 意識不明瞭：実質100%

これを踏まえて回答します
- 新仕様における魔法等の失率は、現行仕様よりマシな値として設定しています
- そもそも朦朧値が現行仕様より蓄積しにくいので、新仕様でプレイする初心者が「ひどく朦朧以上の状態に陥りつつ、HPは十分ある」という状況に至る可能性はかなり稀と思われます
- 新設するであろうアイテム失率ですが、現行仕様での「意識不明瞭」＝新仕様での「ひどく朦朧」でもって、失率を発生させる方針とします
  やや朦朧～朦朧までは失率0%とします
  即ち、「現行仕様における意識不明瞭よりはアイテムが使える可能性があるだけまだマシ (しかも可能性は高い)」として位置づけます
- 巻物は盲目だと読めなくなるので、元々100%確実な危機回避手段とは言えず、失率を設ける方針は続行します
- その他の具体的な内容は引き続き調査・検討します (薬は失率0にする？ 他についても失率が0でないなら何%にする？ 5とか10とか？)

**Comment by Hourier** [comment]
以下の機能はAlpha40に入る予定です
- 朦朧蓄積ランク・朦朧ランクの調整
- 魔道具等の失率

アイテムとは無関係の箇所 (移動・攻撃・回避等)はすぐ拡張するのが困難であることが分かりました
これは第二次実装ということで、Alpha41以降に改めて実施します
